### PR TITLE
Add PyTorch example

### PR DESCRIPTION
**Pull request by htahir1**
## Describe changes
I noticed via Slack that we had no Pytorch example at all. Thought that was a glaring ommision. Tried to conform to the new style with @AlexejPenner but not sure if i managed.

I also took the chance to add a new materializer for torch.utils.dataloader and simplified the old materializer i wrote ages ago.

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [x] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/features/integrations) table.
- [ ] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: examples/pytorch/run.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
+#  Copyright (c) ZenML GmbH 2021. All Rights Reserved.
```
Comment (AlexejPenner) [inline examples/pytorch/run.py:1]:
```suggestion
#  Copyright (c) ZenML GmbH 2022. All Rights Reserved.
```

**Review by AlexejPenner — APPROVED** [review]
Good to go from my side. Feel free to add a generic test to make sure it at least runs through with all steps completed:

Go to zenml/tests/integration/test_examples.py and add a ExampleIntegrationTestConfiguration to the examples list around line 80 :

```python
    ExampleIntegrationTestConfiguration(
        name="pytorch",
        validation_function=generate_basic_validation_function(
            pipeline_name="fashion_mnist_pipeline", step_count=3
        ),
    ),
```

File: examples/pytorch/setup.sh [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+set -Eeo pipefail
+
+pre_run () {
+  pip install torchvision
```
Comment (schustmi) [inline examples/pytorch/setup.sh:6]: Do we need this `torchvision` install just for the dataset?

File: docs/book/features/integrations.md [hunk 3]
```diff
@@ -49,12 +49,12 @@ These are the third-party integrations that ZenML currently supports:
...
 | PyTorch                   | ✅      | Training               |                                                                                               |                                                                                                                                                          |
-| PyTorch Lightning         | ✅      | Training               |                                                                                               |                                                                                                                                                          |
+| PyTorch Lightning         | ✅      | Training               |                                                                                               | [pytorch](https://github.com/zenml-io/zenml/tree/main/examples/pytorch)                                                                                  |
```
Comment (schustmi) [inline docs/book/features/integrations.md:52]: I think this ended up in the wrong line, should be with `pytorch` instead right?

**Review by schustmi — CHANGES_REQUESTED** [review]
Few minor suggestions/questions, otherwise LGTM

**Comment by schustmi** [comment]
@AlexejPenner I think it's also still missing the super quickstart section in the example readme, something to add on this branch or on yours?

File: examples/pytorch/setup.sh (hunk 2 above)
Comment (htahir1) [reply examples/pytorch/setup.sh:6]: Yes, its the easiest way. Any suggestions to the contrary?

File: examples/pytorch/README.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
+# 🔦 Using PyTorch with ZenML
+
+This example demonstrate how we can use ZenML and Torch to build, train, & test ML models.
```
Comment (strickvl) [inline examples/pytorch/README.md:3]:
```suggestion
This example demonstrate how we can use ZenML and PyTorch to build, train, and test ML models.
```

File: examples/pytorch/README.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+[PyTorch](https://pytorch.org/) is an open source machine learning framework that accelerates the path from research prototyping to production deployment.
```
Comment (strickvl) [inline examples/pytorch/README.md:5]:
```suggestion
[PyTorch](https://pytorch.org/) is an open-source machine learning framework that accelerates the path from research prototyping to production deployment.
```

File: examples/pytorch/README.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+With the ZenML pytorch integration, you can pass `torch.nn.Module` and `torch.utils.data.DataLoader` objects through steps as first class citizens. ZenML will automatically make sure
```
Comment (strickvl) [inline examples/pytorch/README.md:7]:
```suggestion
With the ZenML PyTorch integration, you can pass `torch.nn.Module` and `torch.utils.data.DataLoader` objects through steps as first class citizens. ZenML will automatically make sure 
```

File: examples/pytorch/run.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    # Download test data from open datasets.
+    test_data = datasets.FashionMNIST(
+        root="data",
+        train=False,
+        download=True,
+        transform=ToTensor(),
+    )
+    batch_size = 64
+
+    # Create data loaders.
```
Comment (strickvl) [inline examples/pytorch/run.py:70]:
```suggestion
    # Create dataloaders.
```

File: examples/pytorch/run.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+
+# Define the pipeline
+@pipeline(required_integrations=[PYTORCH])
+def fashion_mnist_pipeline(
+    importer,
+    trainer,
+    evaluator,
+):
+    """Link all the steps artifacts together"""
```
Comment (strickvl) [inline examples/pytorch/run.py:133]:
```suggestion
    """Link all the steps and artifacts together"""
```

File: examples/pytorch/setup.sh (hunk 2 above)
Comment (strickvl) [reply examples/pytorch/setup.sh:6]: If we do a `pip install`, this means that `integration uninstall` will not properly clean up and the user will be left with a `torchvision` installed.

Perhaps we have [one of these](https://github.com/zalandoresearch/fashion-mnist#loading-data-with-other-machine-learning-libraries) libraries installed. Maybe grab it via Tensorflow?

File: src/zenml/integrations/pytorch/materializers/pytorch_dataloader_materializer.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    """Materializer to read/write Pytorch dataloaders."""
+
+    ASSOCIATED_TYPES = (DataLoader,)
+    ASSOCIATED_ARTIFACT_TYPES = (DataArtifact,)
+
+    def handle_input(self, data_type: Type[Any]) -> DataLoader:
+        """Reads and returns a PyTorch dataloader.
+
+        Returns:
+            A loaded pytorch dataloader.
```
Comment (strickvl) [inline src/zenml/integrations/pytorch/materializers/pytorch_dataloader_materializer.py:39]:
```suggestion
            A loaded PyTorch dataloader.
```

**Review by strickvl — CHANGES_REQUESTED** [review]
Just some small nits here and there. Otherwise really glad to see this example added.

**Comment by strickvl** [comment]
@htahir1 one other thing I forgot. It's probably a good idea to add this example into our integration test suite like the others.

File: examples/pytorch/setup.sh (hunk 2 above)
Comment (schustmi) [reply examples/pytorch/setup.sh:6]: No better suggestion unfortunately other than getting it from somewhere else (or using a dataset that somehow comes with base `torch` if there are any). @strickvl Requiring tensorflow in our pytorch example might not be optimal either I think

File: examples/pytorch/README.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+This example demonstrate how we can use ZenML and PyTorch to build, train, and test ML models.
+
+[PyTorch](https://pytorch.org/) is an open-source machine learning framework that accelerates the path from research prototyping to production deployment.
+
+With the ZenML PyTorch integration, you can pass `torch.nn.Module` and `torch.utils.data.DataLoader` objects through steps as first class citizens. ZenML will automatically make sure 
+to track and version these objects.
+
+# 🖥 Run it locally
+
+### 📄 Prerequisites
```
Comment (schustmi) [inline examples/pytorch/README.md:12]:
```suggestion

## ⏩ SuperQuick `pytorch` run

If you're really in a hurry and just want to see this example pipeline run
without wanting to fiddle around with all the individual installation and
configuration steps, just run the following:

```shell
zenml example run pytorch
```

## 👣 Step-by-Step

### 📄 Prerequisites
```

Missing quickstart section

**Comment by schustmi** [comment]
> @AlexejPenner I think it's also still missing the super quickstart section in the example readme, something to add on this branch or on yours?

I added this

File: src/zenml/integrations/pytorch/materializers/pytorch_module_materializer.py [hunk 11]
```diff
@@ -54,8 +54,8 @@ def handle_return(self, model: Module) -> None:
         """
         super().handle_return(model)
 
-        # Save entire model to artifact directory,
-        # This is the default behavior for loading model in development phase (training, evaluation)
+        # Save entire model to artifact directory, This is the default behavior
```
Comment (strickvl) [inline src/zenml/integrations/pytorch/materializers/pytorch_module_materializer.py:57]:
```suggestion
        # Save entire model to artifact directory. This is the default behavior
```

**Review by strickvl — APPROVED** [review]
One tiny typo left, but otherwise good to go from my side.

**Pull request merged**